File: Python/300_LongestIncreasingSubsequence.py
class Solution(object):

    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        def bt(nums, idx, curv, lgth, memo):
            if idx == len(nums):    # base 
                return lgth
            elif memo[idx] != 0:
                return memo[idx]
            else:
                mlgth = 0
                for i in range(idx,len(nums)):
                    if curv < nums[i]:
                        mlgth = max(mlgth, bt(nums,i+1,nums[i],lgth+1,memo))
                    
                memo[i] = mlgth

            return mlgth

        memo = [0 for _ in range(len(nums))]

        lgth = 0
        for i in range(len(nums)):
            lgth = max(lgth, bt(nums,i+1,nums[i],1,memo))
        
        return lgth
        
# 0,1,0,3,2,3
# 0,1,3
# 0(4),1(3),2(2),3(1)

# Backtracking algorithm
# time O(N^2)   N is number of nums
# memory O(N) 
# + memoization

File: C/300_LongestIncreasingSubsequence.c
int btLIS(int* nums, int numsSize, int curIdx, int* memo) {
    if (numsSize == curIdx) return 0;  // base case

    if (memo[curIdx] != -1) return memo[curIdx];

    int mx = 1;
    for (int prev = 0; prev < curIdx; prev++) {
        if (nums[prev] < nums[curIdx]) {
            int current = btLIS(nums, numsSize, prev, memo) + 1;
            if (current > mx)
                mx = current;
        }
    }

    memo[curIdx] = mx;

    return mx;
}

int lengthOfLIS(int* nums, int numsSize) {
    int memo[numsSize];

    for (int i=0;i<numsSize;i++){
        memo[i] = -1;
    }

    int max = 1;

    for (int i=1;i<numsSize;i++){
        int lgt = btLIS(nums, numsSize, i, memo);
        if (max < lgt) max = lgt;
    }

    return max;
}

/*
The goal of this algorithm is to find the longest subsequence of increasing elements while preserving their original order in the array. 
I use a backtracking approach to solve this problem.
In this approach, the algorithm traverses the array from the beginning to the end, tracking the length of the subsequence along the way. 
If the next elements is larger than the current elements, the subsequence length is incremented. Otherwise, the algorithm continues to the next element without updating the length. 
To explore all possible combinations of subsequences, the search starts from each element in the array.
*/